File: jambeth/jAmbeth.Util/src/de/osthus/ambeth/util/ImmutableTypeSet.java
package de.osthus.ambeth.util;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Collection;

import de.osthus.ambeth.collections.HashMap;
import de.osthus.ambeth.collections.HashSet;

public final class ImmutableTypeSet
{
	protected static final HashSet<Class<?>> valueTypeSet = new HashSet<Class<?>>(0.5f);

	private static final HashMap<Class<?>, Class<?>> wrapperTypesMap = new HashMap<Class<?>, Class<?>>(0.5f);

	static
	{
		wrapperTypesMap.put(Integer.class, Integer.TYPE);
		wrapperTypesMap.put(Long.class, Long.TYPE);
		wrapperTypesMap.put(Double.class, Double.TYPE);
		wrapperTypesMap.put(Float.class, Float.TYPE);
		wrapperTypesMap.put(Short.class, Short.TYPE);
		wrapperTypesMap.put(Character.class, Character.TYPE);
		wrapperTypesMap.put(Byte.class, Byte.TYPE);
		wrapperTypesMap.put(Boolean.class, Boolean.TYPE);

		valueTypeSet.add(Integer.class);
		valueTypeSet.add(Integer.TYPE);
		valueTypeSet.add(Long.class);
		valueTypeSet.add(Long.TYPE);
		valueTypeSet.add(Double.class);
		valueTypeSet.add(Double.TYPE);
		valueTypeSet.add(Float.class);
		valueTypeSet.add(Float.TYPE);
		valueTypeSet.add(Short.class);
		valueTypeSet.add(Short.TYPE);
		valueTypeSet.add(Character.class);
		valueTypeSet.add(Character.TYPE);
		valueTypeSet.add(Byte.class);
		valueTypeSet.add(Byte.TYPE);
		valueTypeSet.add(Boolean.class);
		valueTypeSet.add(Boolean.TYPE);
		valueTypeSet.add(String.class);
		valueTypeSet.add(Class.class);
		valueTypeSet.add(void.class);
		valueTypeSet.add(BigInteger.class);
		valueTypeSet.add(BigDecimal.class);
	}

	public static void addImmutableTypesTo(Collection<Class<?>> collection)
	{
		collection.addAll(valueTypeSet);
	}

	public static boolean isImmutableType(Class<?> type)
	{
		return type.isPrimitive() || type.isEnum() || valueTypeSet.contains(type) || IImmutableType.class.isAssignableFrom(type);
	}

	private ImmutableTypeSet()
	{
		// Intended blank
	}

	public static Class<?> getUnwrappedType(Class<?> wrapperType)
	{
		return wrapperTypesMap.get(wrapperType);
	}
}

File: ambeth/Ambeth.Util/ambeth/util/ImmutableTypeSet.cs
using System;
using System.Collections.Generic;
using De.Osthus.Ambeth.Collections;

namespace De.Osthus.Ambeth.Util
{
    public class ImmutableTypeSet
    {
	    protected static readonly CHashSet<Type> valueTypeSet = new CHashSet<Type>(0.5f);

        protected static readonly HashMap<Type, Type> wrapperTypesMap = new HashMap<Type, Type>(0.5f);

	    static ImmutableTypeSet()
	    {
            wrapperTypesMap.Put(typeof(Int64?), typeof(Int64));
            wrapperTypesMap.Put(typeof(UInt64?), typeof(UInt64));
            wrapperTypesMap.Put(typeof(Int32?), typeof(Int32));
            wrapperTypesMap.Put(typeof(UInt32?), typeof(UInt32));
            wrapperTypesMap.Put(typeof(Int16?), typeof(Int16));
            wrapperTypesMap.Put(typeof(UInt16?), typeof(UInt16));
            wrapperTypesMap.Put(typeof(Byte?), typeof(Byte));
            wrapperTypesMap.Put(typeof(SByte?), typeof(SByte));
            wrapperTypesMap.Put(typeof(Double?), typeof(Double));
            wrapperTypesMap.Put(typeof(Single?), typeof(Single));
            wrapperTypesMap.Put(typeof(Char?), typeof(Char));
            wrapperTypesMap.Put(typeof(Boolean?), typeof(Boolean));

            valueTypeSet.Add(typeof(Int64));
            valueTypeSet.Add(typeof(Int64?));
            valueTypeSet.Add(typeof(Int32));
            valueTypeSet.Add(typeof(Int32?));
            valueTypeSet.Add(typeof(Int16));
            valueTypeSet.Add(typeof(Int16?));
            valueTypeSet.Add(typeof(UInt64));
            valueTypeSet.Add(typeof(UInt64?));
            valueTypeSet.Add(typeof(UInt32));
            valueTypeSet.Add(typeof(UInt32?));
            valueTypeSet.Add(typeof(UInt16));
            valueTypeSet.Add(typeof(UInt16?));
            valueTypeSet.Add(typeof(Byte));
            valueTypeSet.Add(typeof(Byte?));
            valueTypeSet.Add(typeof(SByte));
            valueTypeSet.Add(typeof(SByte?));
            valueTypeSet.Add(typeof(Char));
            valueTypeSet.Add(typeof(Char?));
            valueTypeSet.Add(typeof(Boolean));
            valueTypeSet.Add(typeof(Boolean?));
            valueTypeSet.Add(typeof(Double));
            valueTypeSet.Add(typeof(Double?));
            valueTypeSet.Add(typeof(Single));
            valueTypeSet.Add(typeof(Single?));
            valueTypeSet.Add(typeof(String));
            valueTypeSet.Add(typeof(Type));
            valueTypeSet.Add(typeof(void));
	    }

	    public static void AddImmutableTypesTo(IList<Type> collection)
	    {
            foreach (Type valueType in valueTypeSet)
            {
		        collection.Add(valueType);
            }
	    }

        public static void AddImmutableTypesTo(IICollection<Type> collection)
        {
            foreach (Type valueType in valueTypeSet)
            {
                collection.Add(valueType);
            }
        }

	    public static bool IsImmutableType(Type type)
	    {
		    return type.IsPrimitive || type.IsValueType || type.IsEnum || valueTypeSet.Contains(type) || "RuntimeType".Equals(type.Name);
	    }

	    private ImmutableTypeSet()
	    {
		    // Intended blank
	    }

        public static Type GetUnwrappedType(Type wrapperType)
	    {
		    return wrapperTypesMap.Get(wrapperType);
	    }
    }
}
